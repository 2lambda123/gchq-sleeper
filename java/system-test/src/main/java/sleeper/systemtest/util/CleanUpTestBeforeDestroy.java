/*
 * Copyright 2022-2023 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.systemtest.util;

import sleeper.configuration.properties.local.LoadLocalProperties;
import sleeper.configuration.properties.table.TableProperties;
import sleeper.status.update.CleanUpBeforeDestroy;
import sleeper.systemtest.SystemTestProperties;

import java.nio.file.Path;
import java.util.List;
import java.util.stream.Collectors;

import static sleeper.systemtest.SystemTestProperty.SYSTEM_TEST_CLUSTER_NAME;

public class CleanUpTestBeforeDestroy {

    private CleanUpTestBeforeDestroy() {
    }

    public static void main(String[] args) {
        if (1 != args.length) {
            throw new IllegalArgumentException("Usage: <configuration directory>");
        }
        Path baseDir = Path.of(args[0]);
        SystemTestProperties instanceProperties = LoadLocalProperties.loadInstancePropertiesFromDirectory(new SystemTestProperties(), baseDir);
        List<TableProperties> tablePropertiesList = LoadLocalProperties
                .loadTablesFromDirectory(instanceProperties, baseDir).collect(Collectors.toList());

        CleanUpBeforeDestroy.cleanUp(instanceProperties, tablePropertiesList,
                List.of(instanceProperties.get(SYSTEM_TEST_CLUSTER_NAME)));
    }
}
