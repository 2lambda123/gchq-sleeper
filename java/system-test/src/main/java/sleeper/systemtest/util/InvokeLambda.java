/*
 * Copyright 2023 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.systemtest.util;

import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;
import sleeper.configuration.properties.InstanceProperty;
import sleeper.systemtest.SystemTestProperties;
import software.amazon.awssdk.core.SdkBytes;
import software.amazon.awssdk.services.lambda.LambdaClient;

import java.io.IOException;

public class InvokeLambda {

    private InvokeLambda() {
    }

    public static void forInstance(String instanceId, InstanceProperty lambdaFunctionProperty) throws IOException {

        AmazonS3 s3Client = AmazonS3ClientBuilder.defaultClient();
        SystemTestProperties systemTestProperties = new SystemTestProperties();
        systemTestProperties.loadFromS3GivenInstanceId(s3Client, instanceId);
        s3Client.shutdown();

        try (LambdaClient lambda = LambdaClient.create()) {
            lambda.invoke(builder -> builder
                    .functionName(systemTestProperties.get(lambdaFunctionProperty))
                    .payload(SdkBytes.fromUtf8String("{}")));
        }
    }
}
