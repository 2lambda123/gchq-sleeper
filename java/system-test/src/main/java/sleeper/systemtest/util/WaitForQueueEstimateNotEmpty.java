/*
 * Copyright 2022-2023 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sleeper.systemtest.util;

import com.amazonaws.services.sqs.AmazonSQS;
import com.amazonaws.services.sqs.model.GetQueueAttributesRequest;
import com.amazonaws.services.sqs.model.GetQueueAttributesResult;
import com.amazonaws.services.sqs.model.QueueAttributeName;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import sleeper.configuration.properties.InstanceProperties;
import sleeper.configuration.properties.InstanceProperty;

public class WaitForQueueEstimateNotEmpty {
    private static final Logger LOGGER = LoggerFactory.getLogger(WaitForQueueEstimateNotEmpty.class);
    private static final long POLL_INTERVAL_MILLIS = 5000;
    private static final int MAX_POLLS = 12;

    private final AmazonSQS sqsClient;
    private final String queueUrl;
    private final PollWithRetries poll = PollWithRetries.intervalAndMaxPolls(POLL_INTERVAL_MILLIS, MAX_POLLS);

    public WaitForQueueEstimateNotEmpty(AmazonSQS sqsClient, InstanceProperties instanceProperties, InstanceProperty queueProperty) {
        this.sqsClient = sqsClient;
        this.queueUrl = instanceProperties.get(queueProperty);
    }

    public void pollUntilFinished() throws InterruptedException {
        LOGGER.info("Waiting for messages on {}", queueUrl);
        poll.pollUntil("queue estimated not empty", this::isEstimateNotEmpty);
    }

    private boolean isEstimateNotEmpty() {
        int approximateMessages = getApproximateMessages();
        LOGGER.info("Approximate number of messages: {}", approximateMessages);
        return approximateMessages > 0;
    }

    private int getApproximateMessages() {
        GetQueueAttributesResult result = sqsClient.getQueueAttributes(new GetQueueAttributesRequest()
                .withQueueUrl(queueUrl)
                .withAttributeNames(QueueAttributeName.ApproximateNumberOfMessages));
        return getInt(result, QueueAttributeName.ApproximateNumberOfMessages);
    }

    private static int getInt(GetQueueAttributesResult result, QueueAttributeName attribute) {
        return Integer.parseInt(result.getAttributes().get(attribute.toString()));
    }
}
