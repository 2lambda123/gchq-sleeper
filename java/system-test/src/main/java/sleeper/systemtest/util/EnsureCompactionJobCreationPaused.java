/*
 * Copyright 2022-2023 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.systemtest.util;

import com.amazonaws.services.cloudwatchevents.AmazonCloudWatchEvents;
import com.amazonaws.services.cloudwatchevents.AmazonCloudWatchEventsClientBuilder;
import com.amazonaws.services.cloudwatchevents.model.DescribeRuleRequest;
import com.amazonaws.services.cloudwatchevents.model.DescribeRuleResult;
import com.amazonaws.services.cloudwatchevents.model.RuleState;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;

import sleeper.configuration.properties.InstanceProperties;
import sleeper.util.ClientUtils;

import java.io.IOException;

import static sleeper.configuration.properties.SystemDefinedInstanceProperty.COMPACTION_JOB_CREATION_CLOUDWATCH_RULE;

public class EnsureCompactionJobCreationPaused {

    private EnsureCompactionJobCreationPaused() {
    }

    public static void main(String[] args) throws IOException {
        if (args.length != 1) {
            System.out.println("Usage: <instance id>");
            return;
        }

        AmazonS3 amazonS3 = AmazonS3ClientBuilder.defaultClient();
        InstanceProperties instanceProperties = ClientUtils.getInstanceProperties(amazonS3, args[0]);
        amazonS3.shutdown();

        AmazonCloudWatchEvents cwClient = AmazonCloudWatchEventsClientBuilder.defaultClient();

        String ruleName = instanceProperties.get(COMPACTION_JOB_CREATION_CLOUDWATCH_RULE);

        DescribeRuleResult result = cwClient.describeRule(new DescribeRuleRequest().withName(ruleName));
        RuleState state = RuleState.fromValue(result.getState());
        if (state != RuleState.DISABLED) {
            System.err.println("Compaction job creation CloudWatch rule is not paused");
            System.exit(1);
        }
    }
}
