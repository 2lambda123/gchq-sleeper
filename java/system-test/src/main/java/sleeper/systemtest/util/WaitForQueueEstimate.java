/*
 * Copyright 2022-2023 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sleeper.systemtest.util;

import com.amazonaws.services.sqs.AmazonSQS;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import sleeper.clients.util.PollWithRetries;
import sleeper.compaction.job.CompactionJobStatusStore;
import sleeper.configuration.properties.InstanceProperties;
import sleeper.configuration.properties.InstanceProperty;
import sleeper.job.common.QueueMessageCount;

import java.util.function.Predicate;

import static sleeper.job.common.QueueMessageCount.withSqsClient;

public class WaitForQueueEstimate {
    private static final Logger LOGGER = LoggerFactory.getLogger(WaitForQueueEstimate.class);

    private final QueueMessageCount.Client queueClient;
    private final Predicate<QueueMessageCount> isFinished;
    private final String description;
    private final String queueUrl;

    public static WaitForQueueEstimate notEmpty(
            QueueMessageCount.Client queueClient, InstanceProperties instanceProperties, InstanceProperty queueProperty) {
        String queueUrl = instanceProperties.get(queueProperty);
        return new WaitForQueueEstimate(queueClient, queueUrl,
                estimate -> estimate.getApproximateNumberOfMessages() > 0,
                "estimate not empty for queue " + queueUrl);
    }

    public static WaitForQueueEstimate isEmpty(
            AmazonSQS sqsClient, InstanceProperties instanceProperties, InstanceProperty queueProperty) {
        String queueUrl = instanceProperties.get(queueProperty);
        return new WaitForQueueEstimate(withSqsClient(sqsClient), queueUrl,
                estimate -> estimate.getApproximateNumberOfMessages() == 0,
                "estimate empty for queue " + queueUrl);
    }

    public static WaitForQueueEstimate containsUnfinishedJobs(
            AmazonSQS sqsClient, InstanceProperties instanceProperties, InstanceProperty queueProperty,
            CompactionJobStatusStore statusStore, String tableName) {
        int unfinished = statusStore.getUnfinishedJobs(tableName).size();
        LOGGER.info("Found {} unfinished compaction jobs", unfinished);
        return new WaitForQueueEstimate(withSqsClient(sqsClient), instanceProperties, queueProperty,
                estimate -> estimate.getApproximateNumberOfMessages() >= unfinished,
                "queue estimate matching unfinished compaction jobs");
    }

    private WaitForQueueEstimate(QueueMessageCount.Client queueClient,
                                 InstanceProperties properties, InstanceProperty queueProperty,
                                 Predicate<QueueMessageCount> isFinished, String description) {
        this(queueClient, properties.get(queueProperty), isFinished, description);
    }

    private WaitForQueueEstimate(QueueMessageCount.Client queueClient, String queueUrl,
                                 Predicate<QueueMessageCount> isFinished, String description) {
        this.queueClient = queueClient;
        this.queueUrl = queueUrl;
        this.isFinished = isFinished;
        this.description = description;
    }

    public void pollUntilFinished(PollWithRetries poll) throws InterruptedException {
        LOGGER.info("Waiting until {}", description);
        poll.pollUntil(description, this::isFinished);
    }

    private boolean isFinished() {
        QueueMessageCount count = queueClient.getQueueMessageCount(queueUrl);
        LOGGER.info("Message count: {}", count);
        return isFinished.test(count);
    }
}
