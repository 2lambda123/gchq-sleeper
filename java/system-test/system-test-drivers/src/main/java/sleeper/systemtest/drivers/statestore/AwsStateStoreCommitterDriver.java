/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.systemtest.drivers.statestore;

import com.amazonaws.services.sqs.AmazonSQS;
import com.amazonaws.services.sqs.model.SendMessageBatchRequest;
import com.amazonaws.services.sqs.model.SendMessageBatchRequestEntry;

import sleeper.core.util.SplitIntoBatches;
import sleeper.systemtest.dsl.instance.SystemTestInstanceContext;
import sleeper.systemtest.dsl.statestore.StateStoreCommitMessage;
import sleeper.systemtest.dsl.statestore.StateStoreCommitterDriver;

import java.util.List;
import java.util.UUID;
import java.util.stream.Stream;

import static java.util.stream.Collectors.toUnmodifiableList;
import static sleeper.configuration.properties.instance.CdkDefinedInstanceProperty.STATESTORE_COMMITTER_QUEUE_URL;

public class AwsStateStoreCommitterDriver implements StateStoreCommitterDriver {

    private final SystemTestInstanceContext instance;
    private final AmazonSQS sqs;

    public AwsStateStoreCommitterDriver(SystemTestInstanceContext instance, AmazonSQS sqs) {
        this.instance = instance;
        this.sqs = sqs;
    }

    @Override
    public void sendCommitMessages(Stream<StateStoreCommitMessage> messages) {
        SplitIntoBatches.inParallelBatchesOf(10, messages, this::sendMessageBatch);
    }

    private void sendMessageBatch(List<StateStoreCommitMessage> batch) {
        sqs.sendMessageBatch(new SendMessageBatchRequest()
                .withQueueUrl(instance.getInstanceProperties().get(STATESTORE_COMMITTER_QUEUE_URL))
                .withEntries(batch.stream()
                        .map(message -> new SendMessageBatchRequestEntry()
                                .withMessageDeduplicationId(UUID.randomUUID().toString())
                                .withId(UUID.randomUUID().toString())
                                .withMessageGroupId(message.getTableId())
                                .withMessageBody(message.getBody()))
                        .collect(toUnmodifiableList())));
    }

}
