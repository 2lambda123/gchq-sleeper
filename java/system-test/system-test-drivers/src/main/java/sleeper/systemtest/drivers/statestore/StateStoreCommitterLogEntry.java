/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.systemtest.drivers.statestore;

import sleeper.systemtest.dsl.statestore.StateStoreCommitSummary;

import java.time.Instant;
import java.util.Objects;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class StateStoreCommitterLogEntry {

    private static final Pattern MESSAGE_PATTERN = Pattern.compile("" +
            "Lambda started at (.+)|" + // Lambda started message type
            "Lambda finished at ([^ ]+) |" + // Lambda finished message type
            "Applied request to table ID ([^ ]+) with type ([^ ]+) at time ([^ ]+)"); // Commit applied message type

    private static class CapturingGroups {
        private static final int START_TIME = 1;
        private static final int FINISH_TIME = 2;
        private static final int TABLE_ID = 3;
        private static final int TYPE = 4;
        private static final int COMMIT_TIME = 5;

        private CapturingGroups() {
        }
    }

    private StateStoreCommitterLogEntry() {
    }

    public static Object readEvent(String message) {
        Matcher matcher = MESSAGE_PATTERN.matcher(message);
        if (!matcher.find()) {
            return null;
        }
        // The pattern can only match one type of log message at a time.
        // Each capturing group will be null unless its message type was matched.
        // We determine which type of message was found based on which capturing group is set.
        String startTime = matcher.group(CapturingGroups.START_TIME);
        if (startTime != null) {
            return new LambdaStarted(Instant.parse(startTime));
        }
        String finishTime = matcher.group(CapturingGroups.FINISH_TIME);
        if (finishTime != null) {
            return new LambdaFinished(Instant.parse(finishTime));
        }
        String tableId = matcher.group(CapturingGroups.TABLE_ID);
        if (tableId != null) {
            String type = matcher.group(CapturingGroups.TYPE);
            String commitTime = matcher.group(CapturingGroups.COMMIT_TIME);
            return new StateStoreCommitSummary(tableId, type, Instant.parse(commitTime));
        }
        return null;
    }

    public static class LambdaStarted {
        private final Instant startTime;

        public LambdaStarted(Instant startTime) {
            this.startTime = startTime;
        }

        public Instant getStartTime() {
            return startTime;
        }

        @Override
        public int hashCode() {
            return Objects.hash(startTime);
        }

        @Override
        public boolean equals(Object obj) {
            if (this == obj) {
                return true;
            }
            if (!(obj instanceof LambdaStarted)) {
                return false;
            }
            LambdaStarted other = (LambdaStarted) obj;
            return Objects.equals(startTime, other.startTime);
        }

        @Override
        public String toString() {
            return "LambdaStarted{startTime=" + startTime + "}";
        }
    }

    public static class LambdaFinished {
        private final Instant finishTime;

        public LambdaFinished(Instant finishTime) {
            this.finishTime = finishTime;
        }

        public Instant getFinishTime() {
            return finishTime;
        }

        @Override
        public int hashCode() {
            return Objects.hash(finishTime);
        }

        @Override
        public boolean equals(Object obj) {
            if (this == obj) {
                return true;
            }
            if (!(obj instanceof LambdaFinished)) {
                return false;
            }
            LambdaFinished other = (LambdaFinished) obj;
            return Objects.equals(finishTime, other.finishTime);
        }

        @Override
        public String toString() {
            return "LambdaFinished{finishTime=" + finishTime + "}";
        }
    }

}
