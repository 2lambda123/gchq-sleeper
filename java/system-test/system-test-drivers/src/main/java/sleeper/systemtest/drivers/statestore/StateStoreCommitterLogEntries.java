/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.systemtest.drivers.statestore;

import sleeper.systemtest.dsl.statestore.StateStoreCommitSummary;
import sleeper.systemtest.dsl.statestore.StateStoreCommitterLogEntry;
import sleeper.systemtest.dsl.statestore.StateStoreCommitterLogs;

import java.util.List;
import java.util.Map;

import static java.util.stream.Collectors.groupingBy;
import static java.util.stream.Collectors.summingInt;

public class StateStoreCommitterLogEntries implements StateStoreCommitterLogs {

    private final List<StateStoreCommitterLogEntry> logs;

    public StateStoreCommitterLogEntries(List<StateStoreCommitterLogEntry> logs) {
        this.logs = logs;
    }

    @Override
    public Map<String, Integer> getNumCommitsByTableId() {
        return logs.stream()
                .filter(entry -> entry instanceof StateStoreCommitSummary)
                .map(entry -> (StateStoreCommitSummary) entry)
                .collect(groupingBy(StateStoreCommitSummary::getTableId, summingInt(commit -> 1)));
    }

}
