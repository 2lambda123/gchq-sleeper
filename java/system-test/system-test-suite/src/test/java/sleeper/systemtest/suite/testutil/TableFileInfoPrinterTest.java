/*
 * Copyright 2022-2023 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sleeper.systemtest.suite.testutil;

import org.junit.jupiter.api.Test;

import sleeper.core.partition.PartitionsBuilder;
import sleeper.core.schema.Schema;
import sleeper.core.schema.type.StringType;
import sleeper.core.statestore.FileInfoFactory;

import java.util.List;

import static org.approvaltests.Approvals.verify;
import static sleeper.core.schema.SchemaTestHelper.schemaWithKey;

public class TableFileInfoPrinterTest {

    private final Schema schema = schemaWithKey("key", new StringType());
    private final PartitionsBuilder partitions = new PartitionsBuilder(schema);

    @Test
    void shouldPrintFilesOnLeaves() {
        partitions.rootFirst("root")
                .splitToNewChildren("root", "L", "R", "row-50")
                .splitToNewChildren("L", "LL", "LR", "row-25")
                .splitToNewChildren("R", "RL", "RR", "row-75")
                .splitToNewChildren("LL", "LLL", "LLR", "row-12")
                .splitToNewChildren("LR", "LRL", "LRR", "row-37")
                .splitToNewChildren("RL", "RLL", "RLR", "row-62")
                .splitToNewChildren("RR", "RRL", "RRR", "row-87");

        FileInfoFactory fileInfoFactory = fileInfoFactory();
        verify(TableFileInfoPrinter.printFiles(partitions.buildTree(), List.of(
                fileInfoFactory.partitionFile("LLL", 12),
                fileInfoFactory.partitionFile("LLR", 13),
                fileInfoFactory.partitionFile("LRL", 12),
                fileInfoFactory.partitionFile("LRR", 13),
                fileInfoFactory.partitionFile("RLL", 12),
                fileInfoFactory.partitionFile("RLR", 13),
                fileInfoFactory.partitionFile("RRL", 12),
                fileInfoFactory.partitionFile("RRR", 13))));
    }

    @Test
    void shouldRenamePartitionsByLocation() {
        partitions.rootFirst("base")
                .splitToNewChildren("base", "l", "r", "row-50")
                .splitToNewChildren("l", "ll", "lr", "row-25")
                .splitToNewChildren("r", "rl", "rr", "row-75")
                .splitToNewChildren("ll", "1", "2", "row-12")
                .splitToNewChildren("lr", "3", "4", "row-37")
                .splitToNewChildren("rl", "5", "6", "row-62")
                .splitToNewChildren("rr", "7", "8", "row-87");

        FileInfoFactory fileInfoFactory = fileInfoFactory();
        verify(TableFileInfoPrinter.printFiles(partitions.buildTree(), List.of(
                fileInfoFactory.partitionFile("1", 12),
                fileInfoFactory.partitionFile("2", 13),
                fileInfoFactory.partitionFile("3", 12),
                fileInfoFactory.partitionFile("4", 13),
                fileInfoFactory.partitionFile("5", 12),
                fileInfoFactory.partitionFile("6", 13),
                fileInfoFactory.partitionFile("7", 12),
                fileInfoFactory.partitionFile("8", 13))));
    }

    private FileInfoFactory fileInfoFactory() {
        return FileInfoFactory.from(partitions.buildTree());
    }
}
