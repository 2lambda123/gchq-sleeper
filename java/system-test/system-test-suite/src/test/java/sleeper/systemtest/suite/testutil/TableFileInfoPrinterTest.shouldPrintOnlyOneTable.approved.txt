One table
Active files:
Partition root: 10 records in file root.parquet

